src/sphmultipcqueue.cpp

```diff
@@ -0,0 +1,599 @@
+				/* revert to GNU TM */
+#endif
+				__transaction_atomic {
+					entryPtr = SPHMPMCQAdvanceTail((SPHLFEntryHeader_t **)&headerBlock->qtail,stride,qlo,qhi);
+				}
+#ifdef __HTM__
+			}
+		}
+#endif
+	} else {
+		debug_printf("%s(%p) type check failed\n",__FUNCTION__,queue);
+	}
+#if 0
+	{
+		SPHLFEntryHeader_t **tail_p = (SPHLFEntryHeader_t **)&headerBlock->qtail;
+		SPHLFEntryHeader_t *entryPtr = *tail_p;
+		sphLFEntry_t entrytemp;
+		entrytemp.idUnit = entryPtr->entryID.idUnit;
+		printf("%-6d: tail = %lx entryPtr = %p allocated = %d valid = %d\n",sphdeGetTID(),headerBlock->qtail,entryPtr,entrytemp.detail.allocated,entrytemp.detail.valid);
+	}
+#endif
+	debug_printf("%-6d: %s = %p\n",sphdeGetTID(),__FUNCTION__,entryPtr);
+	return ((SPHLFEntryDirect_t)entryPtr);
+}
+
+SPHLFEntryDirect_t
+SPHMPSCQGetNextCompleteDirect(SPHMPMCQ_t queue)
+{
+	volatile SPHMPMCQHeader *headerBlock = (SPHMPMCQHeader *) queue;
+	SPHLFEntryHeader_t *entryPtr = 0;
+
+	debug_printf("%-6d: %s NEW\n",sphdeGetTID(),__FUNCTION__);
+
+	if (SOMSASCheckBlockSigAndType ((SASBlockHeader *) headerBlock,SAS_RUNTIME_PCQUEUE_TM)) {
+		const unsigned short stride = headerBlock->default_entry_stride;
+		const longPtr_t qlo = headerBlock->startq;
+		const longPtr_t qhi = headerBlock->endq;
+
+		entryPtr = (SPHLFEntryHeader_t *)headerBlock->qtail;
+		sphLFEntry_t entrytemp;
+		entrytemp.idUnit = entryPtr->entryID.idUnit;
+		if (entrytemp.detail.allocated && entrytemp.detail.valid) {
+			longPtr_t new_tail = ((longPtr_t)entryPtr) + stride;
+			if (new_tail >= qhi)
+				new_tail = qlo;
+			entrytemp.detail.allocated = 0;
+			sas_read_barrier();
+			entryPtr->entryID.idUnit = entrytemp.idUnit;
+			headerBlock->qtail = new_tail;
+		} else
+			entryPtr = 0;
+	} else {
+		debug_printf("%s(%p) type check failed\n",__FUNCTION__,queue);
+	}
+	debug_printf("%-6d: %s = %p\n",sphdeGetTID(),__FUNCTION__,entryPtr);
+	return ((SPHLFEntryDirect_t)entryPtr);
+}
+
+int
+SPHMPMCQEntryDirectIsComplete (SPHLFEntryDirect_t directHandle)
+{
+	SPHLFEntryHeader_t  *entryPtr = (SPHLFEntryHeader_t*)directHandle;
+
+	return (entryPtr->entryID.detail.valid == 1);
+}
+
+int
+SPHMPMCQFreeEntryDirect (SPHMPMCQ_t queue,
+                                     SPHLFEntryDirect_t entry)
+{
+	debug_printf("%-6d: %s(%p,%p) NEW\n",sphdeGetTID(),__FUNCTION__,queue,entry);
+	SPHMPMCQHeader *headerBlock = (SPHMPMCQHeader *) queue;
+	int rc = 0;
+
+	if (SOMSASCheckBlockSigAndType ((SASBlockHeader *) headerBlock,
+	                                  SAS_RUNTIME_PCQUEUE_TM))
+	{
+		SPHLFEntryHeader_t *entryPtr =
+				(SPHLFEntryHeader_t *) entry;
+
+		/* Mark the entry unallocated */
+		entryPtr->entryID.idUnit = 0;
+		rc = 1;
+	} else {
+		debug_printf("SPHMPMCQFreeNextEntry(%p) "
+				"type check failed\n",queue);
+	}
+	debug_printf("%-6d: %s(%p,%p) END\n",sphdeGetTID(),__FUNCTION__,queue,entry);
+	return rc;
+}
+
+int
+SPHMPMCQSetCachePrefetch (SPHMPMCQ_t queue, int prefetch)
+{
+	printf("%s unimplemented\n",__FUNCTION__);
```

Comment:
I do see it called in the tests, both for SPSC and MPMC. However, even those the parameter is evaluated and appropriately stored in "options", it is not used anywhere. There is no real implementation in SPSC, either.  Calling SPHSinglePCQueueSetCachePrefetch has no effect. Perhaps it would be better to remove the API altogether until an implementation (for both) exists?